### DISCUSSION TITLE
i accidentally delete Contributed Scripts 1 page on wiki!

### DISCUSSION THREAD

**Discussion by stansu** [autokey/autokey#651] [General]
is there any backup?
very sorry about this.

**Comment by josephj11** [comment]
Thanks for bringing this to our attention quickly.

I don't know of any backup. There definitely should be some (at least, going forward). However, I'm a git/GitHub dilettante and git is all about versions..., so there may be a way. We couldn't be the first ones to encounter this.

If you could, please ask over [here](https://github.community/t/getting-started-with-the-community-forum/186591) and see if there's a solution.

I posted about this on our [Gitter](https://gitter.im/autokey/autokey) page. You can also discuss this there.

**Comment by Elliria** [comment]
All is not lost and we can recover from it rather quickly.

The best-case scenario would be if someone cloned the wiki. In that case, they could create the **Contributed-Scripts-1** page again using the contents in their clone or they could push the cloned version back onto GitHub.

As an alternative:

I had forked the AutoKey repository, but that doesn't grab a copy of the wiki. If I had known that, I would have intentionally cloned the wiki.

For the record, I now have a current clone of the wiki that I created today and I intend to update it any time I notice that it changed or any time that I change it.

If any of you would like to clone it as well, use this command in any directory you would like to use for holding Git repositories and it will create the hidden ```autokey-wiki``` directory inside of that directory for you with the clones repository in it (note that git has to be installed for this command to work):
```
git clone https://github.com/autokey/autokey.wiki.git
```

Meanwhile:

I was unable to find any instructions for using the GitHub site, itself, to get the data back (gone is gone), but the Wayback Machine website has an archive of the AutoKey wiki. It's not up-to-the-minute current, so we'll be getting an older version of the page, but that's better than nothing in the event that nobody recently cloned the wiki and can restore it.

This is the most recent copy of the **Contributed-Scripts-1** page I was able to find while it was still under the **Scripts---contributed-1** name (I had renamed it later) and its latest revision in that version was in January 15 of 2021: [https://web.archive.org/web/20210604053139/https://github.com/autokey/autokey/wiki/Scripts---contributed-1](https://web.archive.org/web/20210604053139/https://github.com/autokey/autokey/wiki/Scripts---contributed-1)

The **Contributed-Scripts-1** page will need to be recreated with all of the proper Markdown code. I grabbed the source code from the Wayback Machine page and also a plain-text version of what's on that page and will recreate the wiki page from it. It should be back up either in a few hours or tomorrow (I've got a few pressing matters to take care of first).

If you want to poke around in Wayback's archives of the AutoKey wiki for yourself and see if you can find a more recent version, here's all of it: [https://web.archive.org/web/20210501000000*/https://github.com/autokey/autokey/wiki](https://web.archive.org/web/20210501000000*/https://github.com/autokey/autokey/wiki)

**Reply by josephj11** [reply]
@bluedrink9 - I wonder if something like this could be added as a GitHub action. It would obviously have to keep more than one version so it wouldn't just preserve the latest version with possible errors in it.

**Reply by BlueDrink9** [reply]
@Elliria does the cloned wiki have a commit log? `git log`

**Comment by BlueDrink9** [comment]
Reverted. This is fixed

**Reply by Elliria** [reply]
Oh, you are magic. How did you do it?

**Reply by BlueDrink9** [reply]
cloned the git repo you linked, then `git revert HEAD` then git push

**Comment by Elliria** [comment]
Yep. Here it is with the email addresses removed:
```
commit 8947f3fafb494feaba0e9ea987f97b960d2ec48e
Author: stansu <USER@DOMAIN>
Date:   Wed Jan 12 15:46:31 2022 +0800

    Destroyed Contributed Scripts 1 (markdown)

commit 466e298fc2942554ad9114d1f3b3365816610037
Author: Elliria <USER@DOMAIN>
Date:   Thu Dec 23 09:37:33 2021 -0500

    Added a mention of and link to the page with the instructions for installing the beta from Pip in Ubuntu or Kubuntu.

commit b9537b197b8fede60792d075e03161a0744c5e79
Author: Elliria <USER@DOMAIN>
Date:   Thu Dec 23 09:29:26 2021 -0500

    Changed sentence-case to lower-case in all mentions of pip. Fixed a typo in the TOC.

commit d7dd86b5444b42f1f12b977c5097cde471bf0f84
Author: Elliria <USER@DOMAIN>
Date:   Thu Dec 23 09:24:15 2021 -0500

    Added a note about dependencies.
```

**Comment by josephj11** [comment]
You guys are great! Now, post something somewhere that shows the rest of us how to fix it if it ever happens again.

**Comment by kreezxil** [comment]
it's a situation like this that we posited having scripts submitted to a folder in the main repo. We could have a format for them too. Like each script comes with a SCRIPTNAME_readme.md.

For example in that readme
Name of Script:
Description:
Notes:

Then, after that script and it's readme are in the right folder on the git, we would update the wiki to point to it.

----
Yes I didn't discuss any of the logistic involved in this.

**Comment by Elliria** [comment]
Thanks for the explanation. Hopefully we can do this ourselves from now on.

**Comment by Elliria** [comment]
That's a great idea from a development point of view, @kreezxil, but some users might not be willing to learn how it's done, whereas they would have been happy to edit an existing wiki page or add one.

I'm glad that @BlueDrink9 figured out how to quickly revert the change so that those of us who like to mess around on the command line can do this, but it would be even better if GitHub would make that sort of process a part of the online interface so that anybody can do it.

**Comment by kreezxil** [comment]
I don't see how it's hard, what extreme use case scenario are you even
imagining? Back the age of the Amiga computer, we did similar to what I
proposed. You can see examples of this on Aminet now. http://aminet.net/

We could setup one of those github.io pages to read the folder too and it
would assemble a page based on those readme files too.

We won't get there unless we try.

Also editing the wiki is just asking for trouble, if you look carefully,
not many people have edited wiki because it is scary in that way and four
of the people you will find, are me from the last decade. Kreezxil,
Itscool, Bowierocks, and Hypershock, I'm all four of those. Considering how
long I've been doing it, I wonder how many of the other wiki editors are
more than one person from the last decade too?

**Reply by BlueDrink9** [reply]
I can't see how pull requests to a git repo will be more user-friendly and reduce contribution fear more than a wiki. We have no way to know why people aren't editing the wiki, so is suggesting it is fear only speculation?

**Reply by Elliria** [reply]
Yeah, a YAML form would be easy to throw together at this point since we ironed out all the kinks when coming up with the issue report form. YAML forms also can accept specific types of input, with Markdown being one of the types on the list, so it would work nicely and would offer an editor with accessible GUI buttons for those who prefer them.

Editing the wiki may not be asking for trouble now that @BlueDrink9 came up with the steps for recovering from unwanted changes, but I can see how handing the world the keys to the kingdom can go in unpredictable directions.

I'm not sure why GitHub separated the wiki from the repository that it's attached to instead of making it part of it, but they're both repositories, with the only difference being that the wiki doesn't provide GUI access to its repository files.

I may be one of the chameleons who changed names over the years. As for why folks don't contribute, the list is endless. It could be because they're busy, don't know how to contribute, are afraid, aren't aware that they can, aren't interested, or any number of other reasons.

I've wanted to contribute more than I have, because I see spots where the wiki could use some love, but time is my main roadblock. The other is when I see something that needs to be done, but is huge. Then I just get overwhelmed with the sheer number of steps it would take to make it happen, so I either nibble away at it slowly and don't share it in its unfinished state or put it off entirely. It's possible others do that as well.

One of the current huge ones is the shared and example scripts that we're talking about here. Those are currently a mess. They exist on several wiki pages, aren't organized, and have been done in a variety of ways. Reorganizing those in a logical manner is something that needs to be done, but it's a monster task, so I just satisfy myself with sticking my head in the sand whenever I think about it.

Your suggestion of putting them into individual files could solve that organization issue, because then they could be linked to from more than one place without ever being officially organized (and could, thus, be organized in a variety of ways only when they're referenced). The same could be done in the wiki, too, but would create a whole lot of wiki pages for anyone who's looking through them to wade through and make sense of. Those could be handled by choosing a naming convention that makes their purpose clear (like script01, script02, etc.). Your method would remove the scripts from the wiki entirely without significantly increasing the number of wiki pages.

On the other hand, having users submit scripts via a YAML form, although relatively easy for the users, could pose some potential drawbacks. One is that it would put more of a burden on the folks who handle those forms. Each form would need the intervention of Someone In Power to approve it.

Another is that only folks who have turned on specific notifications or folks who make it a point to check the scripts for any they don't recognize would know about the new additions. The same could be said for the wiki, though. If a new page is added and not linked to anywhere, only some people would notice that it's even there or it might go completely unnoticed.

Another is that we tend to make mistakes and like to be able to correct them. When we edit a wiki page and make a mistake, we can edit the page again and fix it immediately. With a YAML form, we'd either have to get in touch with the developers (which we may or may not know how to do) or submit an issue that contains an edit and an explanation for why it's necessary which, again, would require intervention from Someone In Power and wouldn't be immediate.

My intention isn't to argue for one way over another, but rather turn each of them this way and that and see how the light hits them. An interesting experiment might be to try out this other method in a fork, although the wiki can't be forked, so an experiment like that would need to have a new wiki created for it. It's a possibility to try, though, so we can see it in action.

**Comment by kreezxil** [comment]
I wasn't even suggesting pull-requests. just pack up the script/macro with
a readme file to describe it, we could have a form on issue tracker to grab
these, yeah, even to build them. like it could have a box to receive the
script/macro and another set of boxes that prompts the submitter for the
various data we would need to classify it.

On another project I'm involved with, that use github pages, they have js
that reads a folder to generate pages. We could have similar for these
submitted scripts.

The wiki would then point to this github pages for the scripts which then
in turn loads the submitted content.

Note: not all of the logistics was discussed to implement this idea.

**Reply by Elliria** [reply]
The generated pages sound intriguing and seem like they would prevent new scripts from going unnoticed. I assume each script could still be linked to individually elsewhere despite being included in the generated pages, right?

**Comment by josephj11** [comment]
I'll go with the majority on this, but I like it the way it is now - especially since recovery is relatively easy. If there is any fear involved, I would think it would be around being "good enough" to post. I don't think it could be any easier technically than just editing a wiki page and learning a little markdown.

A lot of users are intimidated by developers and the development process. It is pretty involved, but it gets generalized into a reluctance to contribute at all because they're not "experts". All we can do is encourage people to do what they can do and make that as easy as possible.

I also don't want to create a bunch of infrastructure we have to maintain beyond AutoKey itself - although I do really like the new issue reporting form. I'm obviously biased on that one as I helped design it.

**Comment by kreezxil** [comment]
Don't kill this conversation because you agree with 2 people. That's not a
majority. Now if 100's of AK users sounded off in your favor that would be
a majority.

However, kill the conversation, just table it, close the issue, the Wiki
has been fixed, and we all managed to say something about it. So the
thoughts are out there.

**Reply by josephj11** [reply]
Not sure what you mean by "kill the conversation". That's just my opinion and there's no way to "close" a discussion issue even if I wanted to. I'm not even a dev on this project, so I can't implement a change like this either way and I have no "veto" power on anything.

**Comment by kreezxil** [comment]
kill means close
